New gpio bundle to manage led.

pasqualenaclerio opened the pull request:
Signed-off-by: Pasquale Naclerio <pasquale.naclerio@eurotech.com>

Brief description of the PR. [e.g. Added `null` check on `object` to avoid `NullPointerException`]

**Description of the solution adopted:**I use GPIOService to access the GPIO of the device. Use the methods for open, close and set value for access GPIO pin, this methods is defined in LedExample.
I use the LedOptions class to access and modify the LED properties, configure Pin and enable Led.
Inside updated i use a simple if and else to mange the Led, through the methods defined in LedExample.

nacleriopasquale committed:
New gpio bundle to maneg led.

Signed-off-by: Pasquale Naclerio <pasquale.naclerio@eurotech.com>

MMaiero commented on kura/examples/org.eclipse.kura.example.gpio.led/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,12 @@
+Bundle-Version: 1.0.0.qualifier
+Automatic-Module-Name: org.eclipse.kura.example.gpio.led
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Import-Package: org.eclipse.kura.configuration;version="1.2.0",
```
This should be a strict interval [1.2,1.3) as you are implementing a configurableComponent

MMaiero commented on kura/examples/org.eclipse.kura.example.gpio.led/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,12 @@
+Automatic-Module-Name: org.eclipse.kura.example.gpio.led
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Import-Package: org.eclipse.kura.configuration;version="1.2.0",
+ org.eclipse.kura.gpio;version="1.1.0",
```
Here you are using an interface so the interval should be [1.1,2.0)

MMaiero commented on kura/examples/org.eclipse.kura.example.gpio.led/OSGI-INF/metatype/org.eclipse.kura.example.gpio.led.LedExample.xml:
```diff
@@ -0,0 +1,34 @@
+            required="true"
+            default="false"
+            description="Switch on and of the led."/>
+            
```
Maybe you want to make configurable the GPIO pin as well

MMaiero commented on kura/examples/org.eclipse.kura.example.gpio.led/OSGI-INF/metatype/org.eclipse.kura.example.gpio.led.LedExample.xml:
```diff
@@ -0,0 +1,34 @@
+            cardinality="0"
+            required="true"
+            default="false"
+            description="Switch on and of the led."/>
```
Maybe you want to change the description to something like: "Switches the selected GPIO port state"

MMaiero commented on kura/examples/org.eclipse.kura.example.gpio.led/OSGI-INF/metatype/org.eclipse.kura.example.gpio.led.LedExample.xml:
```diff
@@ -0,0 +1,34 @@
+<MetaData xmlns="http://www.osgi.org/xmlns/metatype/v1.2.0" localization="en_us">
+    <OCD id="org.eclipse.kura.example.gpio.led.LedExample"
+        name="LedExample"
+        description="Example of switching on and off the led">
```
"This example allows to configure and manage the state of a LED connected to a configurable GPIO pin"

MMaiero commented on kura/examples/org.eclipse.kura.example.gpio.led/OSGI-INF/metatype/org.eclipse.kura.example.gpio.led.LedExample.xml:
```diff
@@ -0,0 +1,34 @@
+-->
+<MetaData xmlns="http://www.osgi.org/xmlns/metatype/v1.2.0" localization="en_us">
+    <OCD id="org.eclipse.kura.example.gpio.led.LedExample"
+        name="LedExample"
```
Gpio Led Example

MMaiero commented on kura/examples/org.eclipse.kura.example.gpio.led/src/org/eclipse/kura/example/gpio/led/LedExample.java:
```diff
@@ -0,0 +1,93 @@
+
+public class LedExample implements ConfigurableComponent {
+
+    private static final Logger s_logger = LoggerFactory.getLogger(LedExample.class);
```
Please rename s_logger to logger as this is a new "best practice" we are using in the project.

MMaiero commented on kura/examples/org.eclipse.kura.example.gpio.led/src/org/eclipse/kura/example/gpio/led/LedExample.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    protected void activate(ComponentContext componentContext, Map<String, Object> properties) {
+        s_logger.info("Bundle " + APP_ID + " has started with config!");
```
To use better the features of log4j, it's better to change this as:
`logger.info("Bundle {} has started with config!", APP_ID);`

MMaiero replied:
Please do the same also with the deactivate method.

MMaiero commented on kura/examples/org.eclipse.kura.example.gpio.led/src/org/eclipse/kura/example/gpio/led/LedExample.java:
```diff
@@ -0,0 +1,93 @@
+            e.printStackTrace();
+        }
+
+        if (this.options.isEnableLed()) {
```
I believe you could refactor this code snippet to a method and reuse the same code for both switching on and off.

MMaiero reviewed (commented):
Thank you for your PR!

Here few comments that may help you with your first contribution.

nacleriopasquale committed:
In kura/examples/org.eclipse.kura.example.gpio.led/META-INF/MANIFEST.MF i change the interval.
In metatype and LedOprions.java i add the Attribute Definition for to make configurable the GPIO pin.
In LedExample.java i created methods to make the code more functional.

Signed-off-by: Pasquale Naclerio <pasquale.naclerio@eurotech.com>

nacleriopasquale committed:
I simplified the code, i deleted if and else. Now i use a single method incovation.

Signed-off-by: Pasquale Naclerio <pasquale.naclerio@eurotech.com>

MMaiero commented on kura/examples/org.eclipse.kura.example.gpio.led/src/org/eclipse/kura/example/gpio/led/LedExample.java:
```diff
@@ -0,0 +1,101 @@
+ *     Eurotech
+ *******************************************************************************/
+
+package org.eclipse.kura.example.gpio.led;
```
Please apply formatting and cleanup to all java files

MMaiero commented on kura/examples/org.eclipse.kura.example.gpio.led/src/org/eclipse/kura/example/gpio/led/LedExample.java:
```diff
@@ -0,0 +1,101 @@
+    	try {
+            pin.open();
+        } catch (KuraGPIODeviceException | KuraUnavailableDeviceException | IOException e) {
+        	logger.error("Exception GPIOService ", e);
```
Maybe it is better spelled as "GPIOService Exception". Then, please make it a constant in order to not have the same text 3 times around the code.

nacleriopasquale committed:
applay formatting

Signed-off-by: Pasquale Naclerio <pasquale.naclerio@eurotech.com>

nacleriopasquale committed:
format LedOptions

Signed-off-by: pasqualenaclerio <pasquale.naclerio@eurotech.com>
